sympy/ntheory/qs.py

```diff
@@ -0,0 +1,517 @@
+    matrix = []
+    for s_relation in smooth_relations:
+        matrix.append(s_relation[2])
+    return matrix
+
+
+def _gauss_mod_2(A):
+    """Fast gaussian reduction for modulo 2 matrix.
+
+    Parameters:
+    ===========
+
+    A : Matrix
+
+    Examples:
+    ========
+
+    >>> from sympy.ntheory.qs import _gauss_mod_2
+    >>> _gauss_mod_2([[0, 1, 1], [1, 0, 1], [0, 1, 0], [1, 1, 1]])
+    ([[[1, 0, 1], 3]],
+     [True, True, True, False],
+     [[0, 1, 0], [1, 0, 0], [0, 0, 1], [1, 0, 1]])
+
+    References:
+    ===========
+
+    .. [1] A fast algorithm for gaussian elimination over GF(2) and
+    its implementation on the GAPP. Cetin K.Koc, Sarath N.Arachchige"""
+    import copy
+    matrix = copy.deepcopy(A)
+    row = len(matrix)
+    col = len(matrix[0])
+    mark = [False]*row
+    for c in range(col):
+        for r in range(row):
+            if matrix[r][c] == 1:
+                break
+        mark[r] = True
+        for c1 in range(col):
+            if c1 == c:
+                continue
+            if matrix[r][c1] == 1:
+                for r2 in range(row):
+                    matrix[r2][c1] = (matrix[r2][c1] + matrix[r2][c]) % 2
+    dependent_row = []
+    for idx, val in enumerate(mark):
+        if val == False:
+            dependent_row.append([matrix[idx], idx])
+    return dependent_row, mark, matrix
+
+
+def _find_factor(dependent_rows, mark, gauss_matrix, index, smooth_relations, N):
+    """Finds proper factor of N. Here, transform the dependent rows as a
+    combination of independent rows of the gauss_matrix to form the desired
+    relation of the form X**2 = Y**2 modN. After obtaining the desired relation
+    we obtain a proper factor of N by gcd(X - Y, N).
+
+    Parameters:
+    ===========
+
+    dependent_rows : denoted dependent rows in the reduced matrix form
+    mark : boolean array to denoted dependent and independent rows
+    gauss_matrix : Reduced form of the smooth relations matrix
+    index : denoted the index of the dependent_rows
+    smooth_relations : Smooth relations vectors matrix
+    N : Number to be factored
+    """
+    from sympy import integer_nthroot
+    idx_in_smooth = dependent_rows[index][1]
+    independent_u = [smooth_relations[idx_in_smooth][0]]
+    independent_v = [smooth_relations[idx_in_smooth][1]]
+    dept_row = dependent_rows[index][0]
+
+    for idx, val in enumerate(dept_row):
+        if val == 1:
+            for row in range(len(gauss_matrix)):
+                if gauss_matrix[row][idx] == 1 and mark[row] == True:
+                    independent_u.append(smooth_relations[row][0])
+                    independent_v.append(smooth_relations[row][1])
+                    break
+
+    u = 1
+    v = 1
+    for i in independent_u:
+        u *= i
+    for i in independent_v:
+        v *= i
+    #assert u**2 % N == v % N
+    v = integer_nthroot(v, 2)[0]
+    return igcd(u - v, N)
+
+
+def qs(N, prime_bound, M, ERROR_TERM=25, threshold=10, seed=1234):
```

Comment:
Can the threshold  parameter be removed, in the paper  it was given that the threshold of 5% of total primes in factorbase is sufficient.

Reply:
What is its significance?

Reply:
The threshold representes extra smooth relations because most probably there will be some redundant/dependent relations , so by taking some extra relations this increases the probability of getting sufficient independent relations. In the          paper it was given that threshold = 5% of facotr_base is a good approximation.